pkg/auth/client/impersonate.go

```diff
@@ -72,3 +62,51 @@ func NewImpersonatingKubernetesClientset(user user.Info, config restclient.Confi
 	impersonatingConfig := NewImpersonatingConfig(user, config)
 	return kclientset.NewForConfig(&impersonatingConfig)
 }
+
+// impersonatingRESTClient sets impersonating user, groups, and scopes headers per request
+type impersonatingRESTClient struct {
+	user     user.Info
+	delegate restclient.Interface
+}
+
+func NewImpersonatingRESTClient(user user.Info, client restclient.Interface) restclient.Interface {
+	return &impersonatingRESTClient{user: user, delegate: client}
+}
+
+// Verb does the impersonation per request by setting the proper headers
+func (c impersonatingRESTClient) Verb(verb string) *restclient.Request {
```

Comment:
this should work fine, and avoids assuptions about the implementations of the delegate's implementations of Post()/Put(), etc:

```go
func (c *impersonatingRESTClient) impersonate(*restclient.Request) *restclient.Request {
  req.SetHeader(authenticationapi.ImpersonateUserHeader, c.user.GetName())
  req.SetHeader(authenticationapi.ImpersonateGroupHeader, c.user.GetGroups()...)
  req.SetHeader(authenticationapi.ImpersonateUserScopeHeader, c.user.GetExtra()[authorizationapi.ScopesKey]...)
  return req
}

func (c *impersonatingRESTClient) Verb(verb string) *restclient.Request {
  return c.impersonate(c.delegate.Verb(verb))
}

func (c impersonatingRESTClient) Post() *restclient.Request {
  return c.impersonate(c.delegate.Post())
}

...
```